scripts/pages/zoning/zoning.ts

```diff
@@ -559,21 +559,37 @@ class ZoneMenuHandler {
 
 	deleteRegion() {
 		if (this.selectedZone.zone?.regions?.length === 1) {
-			$.Warning("Deleted the last Region in this zone!!!");
-			const trackIndex =
-				this.selectedZone.track === this.mapZoneData.tracks.main
-					? 0
-					: this.mapZoneData.tracks.bonuses.indexOf(this.selectedZone.track) + 1;
-			const segmentIndex = this.selectedZone.track.zones.segments.indexOf(this.selectedZone.segment);
-
-			this.deleteSelection();
-			const childContainer = this.panels.trackList.GetChild(trackIndex)?.FindChildTraverse('ChildContainer');
-			const segmentPanel = childContainer?.FindChildTraverse('SegmentContainer')?.GetChild(segmentIndex);
-			const selectButton = segmentPanel.FindChildTraverse<ToggleButton>('SelectButton');
-			const deleteButton = segmentPanel.FindChildTraverse<Button>('DeleteButton');
-
-			selectButton?.SetSelected(true);
-			this.updateSelection(this.selectedZone.track, this.selectedZone.segment, null, deleteButton);
+			UiToolkitAPI.ShowGenericPopupTwoOptionsBgStyle(
+				$.Localize('#Zoning_Delete'),
+				$.Localize('#Zoning_LastRegion_Message'),
+				'warning-popup',
+				$.Localize('#Zoning_Delete'),
+				() => {
+					const trackIndex =
+						this.selectedZone.track === this.mapZoneData.tracks.main
+							? 0
+							: this.mapZoneData.tracks.bonuses.indexOf(this.selectedZone.track) + 1;
+					const segmentIndex = this.selectedZone.track.zones.segments.indexOf(this.selectedZone.segment);
+
+					this.deleteSelection();
+					const childContainer = this.panels.trackList
+						.GetChild(trackIndex)
+						?.FindChildTraverse('ChildContainer');
+					const segmentPanel = childContainer?.FindChildTraverse('SegmentContainer')?.GetChild(segmentIndex);
+					const selectButton = segmentPanel.FindChildTraverse<ToggleButton>('SelectButton');
+					const deleteButton = segmentPanel.FindChildTraverse<Button>('DeleteButton');
+
+					selectButton?.SetSelected(true);
+					this.updateSelection(this.selectedZone.track, this.selectedZone.segment, null, deleteButton);
+				},
+				$.Localize('#Zoning_Recreate'),
+				() => {
+					this.panels.zoningMenu.createRegion(
+						!this.selectedZone.segment?.checkpoints.indexOf(this.selectedZone.zone)
```

Comment:
This should also probably use `includes()` instead of `indexOf()`

Reply:
I think this is the same, we want to know if the zone is the first element in the `checkpoints` array.

I can add the elvis here, but I also kind of want to know if a segment ever exists without the `checkpoints` property (it shouldn't be possible unless a user manually edits the json file).

Reply:
Panzer is right about the `?.` here. In JS `?.` doesn't short-circuit, so if `this.selectedZone.segment` is `null` or `undefined`, `this.selectedZone.segment?.` will evaluated to `undefined` and throw with "trying to access properly "checkpoints" of undefined`.

And yeah the `indexOf` thing still applies, you should make the comparison explicit. So
```js
this.panels.zoningMenu.createRegion(
  this.selectedZone.segment?.checkpoints?.indexOf?.(this.selectedZone.zone) === 0
```

The elvis before the `?.indexOf` looks a bit silly, but think of the `indexOf` function as just being a property of the array (well, array prototype).

Reply:
Incidentally this is why I want to enable TypeScript strict mode, it would've flagged this as being unsafe undefined/null handling.